File: util/cardinaldirection_test.go
package util

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestDirectionTranslate(t *testing.T) {
	// main 4 directions
	{
		newRow, newCol := North.Translate(0, 0)
		assert.Equal(t, -1, newRow)
		assert.Equal(t, 0, newCol)
	}
	{
		newRow, newCol := East.Translate(0, 0)
		assert.Equal(t, 0, newRow)
		assert.Equal(t, 1, newCol)
	}
	{
		newRow, newCol := South.Translate(0, 0)
		assert.Equal(t, 1, newRow)
		assert.Equal(t, 0, newCol)
	}
	{
		newRow, newCol := West.Translate(0, 0)
		assert.Equal(t, 0, newRow)
		assert.Equal(t, -1, newCol)
	}
	// diagonals
	{
		newRow, newCol := NorthEast.Translate(0, 0)
		assert.Equal(t, -1, newRow)
		assert.Equal(t, 1, newCol)
	}
	{
		newRow, newCol := SouthEast.Translate(0, 0)
		assert.Equal(t, 1, newRow)
		assert.Equal(t, 1, newCol)
	}
	{
		newRow, newCol := SouthWest.Translate(0, 0)
		assert.Equal(t, 1, newRow)
		assert.Equal(t, -1, newCol)
	}
	{
		newRow, newCol := NorthWest.Translate(0, 0)
		assert.Equal(t, -1, newRow)
		assert.Equal(t, -1, newCol)
	}
}

func TestString(t *testing.T) {
	assert.Equal(t, "North", North.String())
	assert.Equal(t, "NorthWest", NorthWest.String())
	assert.Equal(t, "South", South.String())
	assert.Equal(t, "SouthEast", SouthEast.String())
}

File: pyutil/cardinal_direction_test.py
from pyutil.cardinal_direction import AllDirections, Direction, PrimaryFourDirections


def test_walk_forward():
    assert Direction.East.walk_forward(0, 0) == (0, 1)
    assert Direction.SouthEast.walk_forward(0, 0) == (1, 1)
    assert Direction.NorthWest.walk_forward(0, 0) == (-1, -1)


def test_walk_backward():
    assert Direction.East.walk_backward(0, 0) == (0, -1)
    assert Direction.SouthEast.walk_backward(0, 0) == (-1, -1)
    assert Direction.NorthWest.walk_backward(0, 0) == (1, 1)


def test_turn():
    assert Direction.North.turn_left() == Direction.West
    assert Direction.North.turn_right() == Direction.East


def test_directions():
    assert PrimaryFourDirections == [
        Direction.North,
        Direction.East,
        Direction.South,
        Direction.West,
    ]
    assert AllDirections == [
        Direction.North,
        Direction.NorthEast,
        Direction.East,
        Direction.SouthEast,
        Direction.South,
        Direction.SouthWest,
        Direction.West,
        Direction.NorthWest,
    ]
